File: 2020/23/main.java
import java.lang.StringBuilder;
import java.util.ArrayList;
import java.util.Arrays;

import tcollier.Executor;
import tcollier.Solution;

class Game {
  private int numRounds;
  private int[] cups;
  private int head;

  public Game(ArrayList<Integer> labels, int numCups, int numRounds) {
    this.numRounds = numRounds;
    this.cups = new int[numCups];

    if (numCups > labels.size()) {
      head = numCups - 1;
    } else {
      head = labels.get(labels.size() - 1) - 1;
    }

    int curr = head;
    int i = 0;
    for (; i < labels.size(); i++) {
      cups[curr] = labels.get(i) - 1;
      curr = labels.get(i) - 1;
    }
    for (; i < numCups; i++) {
      cups[curr] = i;
      curr = i;
    }
  }

  public void play() {
    for (int i = 0; i < numRounds; i++) {
      move();
    }
  }

  private void move() {
    int cup1 = cups[head];
    int cup2 = cups[cup1];
    int cup3 = cups[cup2];
    int cup4 = cups[cup3];
    int[] pickups = {cup2, cup3, cup4};
    cups[cup1] = cups[cup4];

    int dest = cups[head] - 1;
    if (dest < 0) {
        dest = cups.length - 1;
    }
    while (dest == cup2 || dest == cup3 || dest == cup4) {
      dest -= 1;
      if (dest < 0) {
        dest = cups.length - 1;
      }
    }

    int tmp = cups[dest];
    cups[dest] = cup2;
    cups[cup2] = cup3;
    cups[cup3] = cup4;
    cups[cup4] = tmp;

    if (dest == head) {
      head = cup4;
    }

    head = cups[head];
  }

  public int labelAt(int index) {
    return cups[index] + 1;
  }
}

class Day23Solution implements Solution<Integer> {
  public String part1Answer(ArrayList<Integer> labels) {
    Game game = new Game(labels, 9, 100);
    game.play();
    StringBuilder str = new StringBuilder();
    int curr = game.labelAt(0);
    while (curr != 1) {
      str.append(curr);
      curr = game.labelAt(curr - 1);
    }
    return str.toString();
  }

  public String part2Answer(ArrayList<Integer> labels) {
    Game game = new Game(labels, 1000000, 10000000);
    game.play();
    return String.valueOf((long)game.labelAt(0) * (long)game.labelAt(game.labelAt(0) - 1));
  }
}

class Main {
  private static final int[] rawLabels = {1, 9, 8, 7, 5, 3, 4, 6, 2};

  public static void main(String[] args) {
    ArrayList<Integer> labels = new ArrayList<Integer>();
    for (int i = 0; i < rawLabels.length; i++) {
      labels.add(rawLabels[i]);
    }
    Executor<Integer> executor = new Executor<Integer>(new Day23Solution(), labels);
    executor.run(args);
  }
}

File: 2020/23/main.py
import sys

from lib.executor import Executor


def generate_circle(labels, num_cups):
    indexes = [l - 1 for l in labels]
    cups = [None] * num_cups
    if num_cups > len(labels):
        head = num_cups - 1
    else:
        head = indexes[-1]
    curr = head
    for index in indexes:
        cups[curr] = index
        curr = index
    for index in range(len(indexes), num_cups):
        cups[curr] = index
        curr = index
    return cups, head


def move(cups, head, move_num):
    cup1 = cups[head]
    cup2 = cups[cup1]
    cup3 = cups[cup2]
    cup4 = cups[cup3]
    pickups = set([cup2, cup3, cup4])
    cups[cup1] = cups[cup4]

    dest = cups[head] - 1
    if dest < 0:
        dest = len(cups) - 1
    while dest in pickups:
        dest -= 1
        if dest < 0:
            dest = len(cups) - 1

    tmp = cups[dest]
    cups[dest] = cup2
    cups[cup2] = cup3
    cups[cup3] = cup4
    cups[cup4] = tmp

    if dest == head:
        head = cup4
    return head


def output_part1(cups, _):
    labels = []
    curr = cups[0]
    while curr != 0:
        labels.append(str(curr + 1))
        curr = cups[curr]
    return "".join(labels)


def output_part2(cups, head):
    return (cups[0] + 1) * (cups[cups[0]] + 1)


def play(labels, num_cups, num_moves):
    cups, head = generate_circle(labels, num_cups)
    for move_num in range(num_moves):
        head = move(cups, head, move_num + 1)
        head = cups[head]
    return cups, head


def part1_solution(labels):
    cups, head = play(labels, num_cups=9, num_moves=100)
    return output_part1(cups, head)


def part2_solution(labels):
    cups, head = play(labels, num_cups=1000000, num_moves=10000000)
    return output_part2(cups, head)


executor = Executor([1, 9, 8, 7, 5, 3, 4, 6, 2], part1_solution, part2_solution)
executor(sys.argv)
